### PR TITLE
Dyn register rust 02

### PR DESCRIPTION
**Pull request by chifflier**
Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [x] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [x] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [ ] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
None

Describe changes:
- This PR simplifies the registration of parsers, based on the existing method (it calls the same functions in the same way). If it works and is satisfying, it could be extended to all parsers. parsers, based on the existing C method (it calls the same functions in the same way). If it works and is satisfying, it could be extended to all parsers.
- The registration code is factorized into a generic function, taking a structure as input. The structure defines a parser, with all required functions (and optional ones)
- Using this, writing a Rust parser does not require to write C code anymore
- As a first step, it uses the ALPROTO constants from the C code. Adding a dynamic registration method is possible, but would impact much more code. That can be another PR.
- I updated only the NTP parser to show the features, but all Rust parsers can be updated easily after.

Changes since #2951
- Parser registration moved to C
- Parser structure now includes all possible functions (including GetFiles, GetTxMpmID, etc.)

### CODE REVIEW

File: src/app-layer-register.h [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+ * \file
+ *
+ * \author Pierre Chifflier <chifflier@wzdftpd.net>
+ */
+
+#ifndef __APP_LAYER_REGISTER_H__
+#define __APP_LAYER_REGISTER_H__
+
+typedef struct AppLayerParser {
+    const char *Name;
```
Comment (victorjulien) [inline src/app-layer-register.h]: in C our style is to have Funcs and FuncPtrs like this, but variables_like_this. This includes struct members.

File: src/app-layer-parser.h [hunk 2]
```diff
@@ -103,6 +103,12 @@ void AppLayerParserThreadCtxFree(AppLayerParserThreadCtx *tctx);
 int AppLayerParserConfParserEnabled(const char *ipproto,
                                     const char *alproto_name);
 
+/** \brief Prototype for parsing functions */
+typedef int (*ParserFPtr)(Flow *f, void *protocol_state,
```
Comment (victorjulien) [inline src/app-layer-parser.h]: As this is now public I like it to be prefixed with AppLayer.

Comment (chifflier) [reply src/app-layer-parser.h]: Should I use something like `AppLayerParserFPtr` ?
note: there is `ProbingParserFPtr` in `src/app-layer-detect-proto.h`

Comment (chifflier) [reply src/app-layer-parser.h]: Changed to `AppLayerParserFPtr`

File: src/app-layer-register.h (hunk 1 above)
Comment (chifflier) [reply src/app-layer-register.h]: Ok, updated

**Comment by chifflier** [comment]
Replaced by #2955

**Pull request closed without merging**